### ISSUE TITLE
[Question] Model converted using TensorRT is slower than native Pytorch

### ISSUE THREAD

**Issue by AvivSham** [pytorch/TensorRT#2724]
Hi All,
We try to run `resent18` model faster than just running the torchvision version on GPU, therefore we planned to convert and quantize the model using TensorRT. However, we did not witness a performance boost after the conversion. 
We tried to play with the `ir` mode using both `torch_compile` and `dynamo` in addition we tried varying values of `optimization_level` which also did not help. 
Adding here code snip:
```python
import logging
import time

import torch_tensorrt
import torchvision
import torch
from torch.utils.data import DataLoader

from src.utils.utils import set_logger

set_logger()


@torch.no_grad()
def benchmark(model, inputs):
    times = list()
    for i in range(100):
        t = time.time()
        model(inputs)
        torch.cuda.synchronize()
        times.append(time.time() - t)
    return sum(times) / len(times), times


if __name__ == '__main__':
    # dataset = torchvision.datasets.STL10(
    #     root='/tmp/data',
    #     split='train',
    #     download=True,
    #     transform=torchvision.transforms.ToTensor()
    # )
    # loader = DataLoader(dataset, batch_size=2)
    bs = 128
    dummy_input = torch.rand(bs, 3, 96, 96).cuda()
    model = torchvision.models.resnet18(pretrained=True)
    model.fc = torch.nn.Linear(512, 10)  # Change the output layer to have 10 classes
    model.cuda()
    model.eval()

    ir_mode = "dynamo"
    # ir_mode = "torch_compile"

    trt_mod = torch_tensorrt.compile(
        model,
        ir=ir_mode,
        inputs=[torch_tensorrt.Input((bs, 3, 96, 96))],
        enabled_precisions={torch.float32},
        device=torch.device('cuda:0'),
        optimization_level=5,
    )
    avg_time, times = benchmark(model, dummy_input)
    logging.info(f"Model pytorch 32fp: {avg_time}")

    avg_time, times = benchmark(trt_mod, dummy_input)
    logging.info(f"Model compiled to TensorRT 32fp: {avg_time}")

    avg_time, times = benchmark(model.half(), dummy_input.half())
    logging.info(f"Model 16fp: {avg_time}")


    trt_mod = torch_tensorrt.compile(
        model.half(),
        ir=ir_mode,
        inputs=[torch_tensorrt.Input((bs, 3, 96, 96), dtype=torch.half)],
        enabled_precisions={torch.float16},
        device=torch.device('cuda:0'),
        optimization_level=5,
    )
    avg_time, times = benchmark(trt_mod, dummy_input.half())
    logging.info(f"Model compiled to TensorRT 16fp: {avg_time}")

```

**Adding Logs for running with `dynamo`**:
```
torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
  _torch_pytree._register_pytree_node(
/opt/conda/envs/faster-whisper/lib/python3.9/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/opt/conda/envs/faster-whisper/lib/python3.9/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=ResNet18_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet18_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)
INFO:torch_tensorrt.dynamo._compiler:Compilation Settings: CompilationSettings(precision=torch.float32, debug=False, workspace_size=0, min_block_size=5, torch_executed_ops=set(), pass_through_build_failures=False, max_aux_streams=None, version_compatible=False, optimization_level=5, use_python_runtime=False, truncate_long_and_double=False, use_fast_partitioner=True, enable_experimental_decompositions=False, device=Device(type=DeviceType.GPU, gpu_id=0), require_full_compilation=False, disable_tf32=False, sparse_weights=False, refit=False, engine_capability=<EngineCapability.DEFAULT: 0>, num_avg_timing_iters=1, dla_sram_size=1048576, dla_local_dram_size=1073741824, dla_global_dram_size=536870912, output_format='exported_program')

INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:TRT INetwork construction elapsed time: 0:00:00.304469
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:Using optimization level 5
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:Build TRT engine elapsed time: 0:00:17.935200
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:TRT Engine uses: 113246208 bytes of Memory
/opt/conda/envs/faster-whisper/lib/python3.9/site-packages/torch/_export/exported_program.py:333: UserWarning: Unable to execute the generated python source code from the graph. The graph module will no longer be directly callable, but you can still run the ExportedProgram, and if needed, you can run the graph module eagerly using torch.fx.Interpreter.
  warnings.warn(
INFO:root:Model pytorch 32fp: 0.021189916133880615
INFO:root:Model compiled to TensorRT 32fp: 0.02402569055557251
INFO:root:Model 16fp: 0.01123286247253418
INFO:torch_tensorrt.dynamo._compiler:Compilation Settings: CompilationSettings(precision=torch.float16, debug=False, workspace_size=0, min_block_size=5, torch_executed_ops=set(), pass_through_build_failures=False, max_aux_streams=None, version_compatible=False, optimization_level=5, use_python_runtime=False, truncate_long_and_double=False, use_fast_partitioner=True, enable_experimental_decompositions=False, device=Device(type=DeviceType.GPU, gpu_id=0), require_full_compilation=False, disable_tf32=False, sparse_weights=False, refit=False, engine_capability=<EngineCapability.DEFAULT: 0>, num_avg_timing_iters=1, dla_sram_size=1048576, dla_local_dram_size=1073741824, dla_global_dram_size=536870912, output_format='exported_program')

INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:TRT INetwork construction elapsed time: 0:00:00.251508
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:Using optimization level 5
[04/03/2024-18:24:49] [TRT] [W] TensorRT encountered issues when converting weights between types and that could affect accuracy.
[04/03/2024-18:24:49] [TRT] [W] If this is not the desired behavior, please modify the weights or retrain with regularization to adjust the magnitude of the weights.
[04/03/2024-18:24:49] [TRT] [W] Check verbose logs for the list of affected weights.
[04/03/2024-18:24:49] [TRT] [W] - 1 weights are affected by this issue: Detected subnormal FP16 values.
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:Build TRT engine elapsed time: 0:00:40.797937
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:TRT Engine uses: 56623104 bytes of Memory
/opt/conda/envs/faster-whisper/lib/python3.9/site-packages/torch/_export/exported_program.py:333: UserWarning: Unable to execute the generated python source code from the graph. The graph module will no longer be directly callable, but you can still run the ExportedProgram, and if needed, you can run the graph module eagerly using torch.fx.Interpreter.
  warnings.warn(
INFO:root:Model compiled to TensorRT 16fp: 0.0108557391166687
```
Most of the time, this test yields on/sub par results compared with the native pytorch model without a significant boost as expected. 

## Environment
**We are using `g4dn` aws machine with the following env**:
 - PyTorch Version: `2.2.0`
 - CPU Architecture: `x86_64`
 - OS (e.g., Linux): `Linux`
 - How you installed PyTorch: `pip`
 - Python version: `3.9.16`
 - CUDA version: `11.8`
 - GPU models and configuration: `Tesla T4`

What are we missing here? How can we make `resnet18` model run faster (and in general for any pytorch model)?

@avivnavon

**Comment by gs-olive** [comment]
Hello - here are a few things to try:

- For the timing, consider using `time.time_ns()`, then divide by 1e9 instead of `time.time()`, for timing precision
- In the benchmarks (especially when using `ir="torch_compile"`, run the model a few times before benchmarking for warmup, as in:
```python
    # Warmup model
    for _ in range(10):
        model(inputs)
    torch.cuda.synchronize()
    
    ...

    # Benchmark
```
- Try `ir="torch_compile"`
- Try omitting `optimization_level=5` to see the performance with the default optimization level


Potentially related: #2720

**Comment by AvivSham** [comment]
Hi @gs-olive 
Thanks for the tips! 
Here are the logs after using `torch_compile` and omitting `optimization_level`:
```
/opt/conda/envs/faster-whisper/lib/python3.9/site-packages/transformers/utils/generic.py:441: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
  _torch_pytree._register_pytree_node(
/opt/conda/envs/faster-whisper/lib/python3.9/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/opt/conda/envs/faster-whisper/lib/python3.9/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=ResNet18_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet18_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)
INFO:root:Model pytorch 32fp: 0.019387660026550294
INFO:torch_tensorrt.dynamo.utils:Using Default Torch-TRT Runtime (as requested by user)
INFO:torch_tensorrt.dynamo.utils:Compilation Settings: CompilationSettings(precision=torch.float32, debug=False, workspace_size=0, min_block_size=5, torch_executed_ops=set(), pass_through_build_failures=False, max_aux_streams=None, version_compatible=False, optimization_level=None, use_python_runtime=False, truncate_long_and_double=False, use_fast_partitioner=True, enable_experimental_decompositions=False, device=Device(type=DeviceType.GPU, gpu_id=0), require_full_compilation=False, disable_tf32=False, sparse_weights=False, refit=False, engine_capability=<EngineCapability.DEFAULT: 0>, num_avg_timing_iters=1, dla_sram_size=1048576, dla_local_dram_size=1073741824, dla_global_dram_size=536870912, output_format='exported_program')

INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:TRT INetwork construction elapsed time: 0:00:00.412299
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:Build TRT engine elapsed time: 0:00:05.120540
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:TRT Engine uses: 35389440 bytes of Memory
INFO:root:Model compiled to TensorRT 32fp: 0.0071489191055297855
INFO:root:Model 16fp: 0.0037212777137756348
INFO:torch_tensorrt.dynamo.utils:Using Default Torch-TRT Runtime (as requested by user)
INFO:torch_tensorrt.dynamo.utils:Compilation Settings: CompilationSettings(precision=torch.float16, debug=False, workspace_size=0, min_block_size=5, torch_executed_ops=set(), pass_through_build_failures=False, max_aux_streams=None, version_compatible=False, optimization_level=None, use_python_runtime=False, truncate_long_and_double=False, use_fast_partitioner=True, enable_experimental_decompositions=False, device=Device(type=DeviceType.GPU, gpu_id=0), require_full_compilation=False, disable_tf32=False, sparse_weights=False, refit=False, engine_capability=<EngineCapability.DEFAULT: 0>, num_avg_timing_iters=1, dla_sram_size=1048576, dla_local_dram_size=1073741824, dla_global_dram_size=536870912, output_format='exported_program')

INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:TRT INetwork construction elapsed time: 0:00:00.263242
[04/07/2024-09:10:35] [TRT] [W] TensorRT encountered issues when converting weights between types and that could affect accuracy.
[04/07/2024-09:10:35] [TRT] [W] If this is not the desired behavior, please modify the weights or retrain with regularization to adjust the magnitude of the weights.
[04/07/2024-09:10:35] [TRT] [W] Check verbose logs for the list of affected weights.
[04/07/2024-09:10:35] [TRT] [W] - 1 weights are affected by this issue: Detected subnormal FP16 values.
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:Build TRT engine elapsed time: 0:00:15.061290
INFO:torch_tensorrt.dynamo.conversion._TRTInterpreter:TRT Engine uses: 17694720 bytes of Memory
INFO:root:Model compiled to TensorRT 16fp: 0.0017661261558532715
```
This specific run is with `batch_size=40` I tried other sizes as well, and in all of them the compiled model was faster by `~2X`.
@gs-olive do you have more tips to make the model even faster? I want to squeeze as make as possible.

**Comment by gs-olive** [comment]
Hello - some other things to try are to modify the `workspace_size` parameter and to use Nsight Systems to profile runs, to identify which layers/components in the model are taking the longest. This can help with further identifying performance gaps and opportunities

**Comment by gs-olive** [comment]
Generally, the default selection for `workspace` is a good choice. In this case, some values to try might be the total GPU memory, and different percentages of the total to see if the runtimes vary.

`workspace` corresponds to the amount of space allowed to TensorRT to build and run inference with the particular model. Varying the workspace can enable selection of different performance tactics, so this might yield some change in perf.

One other thing to try is wrapping the compilation and inference in `with torch.no_grad()`, to ensure any parts running in Torch are not tracking gradient information.
